### PR TITLE
Implementing Tests

### PR DESCRIPTION
**Pull request by syed-ods**
Please keep this PR open until all tests have been added and approved.

Unit tests:
1. Check if the top level redirect page exists. 
2. The main English template was generated. 
3. The main French template was generated.
4. The sitemap template was generated.

**Integration tests:**
1. Check if the core `ontario-frontend` NPM package exists.
2. If the user opted for ES-Lint, the config package was installed.
3. If the user opted for Prettier, the config package was installed.

**E2E tests:**
1. Test if the user clicks the French and English buttons, the page is populated with appropriate language content/data.
2. Test if the search input and button are working, and the search returns some results. (Integration + E2E)
- Check if search exists
- Make sure typing a search term and clicking the search button redirects you to the   search URL.

**Test Prototypes:**
https://github.com/syed-ods/on-frontend-test-prototypes

<hr>

#### Unit Tests "{new-project}/src/tests/_unit/app.spec.js"
- The user can create a new project using `ontario-create-app`. They should then be able to run `npm test` or `npm run test` from their project.
- The assertion test to check the main index page (dist folder) should fail with the appropriate user feedback. This test should pass after the user builds or serves the app.
- The tests should capture the EN and FR page names via the CLI prompt. Running `npm test` should pass these two tests.
- If the user changes these file names for EN or FR pages, the tests should fail prompting the user to edit that information in the appropriate test files.
- The last unit test for checking sitemap should also pass as this file will be created with every new project.

#### Integration Tests "{new-project}/src/tests/integration/packages.spec.js"
- Running `npm test` from a newly created project should run the integration tests.
- The file `src/tests/integration/packages.spec.js` should capture the ESLint and Prettier user selection.
- If the user selected prettier or eslint, uninstalling any of their associated or core packages should fail the tests for optional packages.
- If the core package `ontario-frontend` is uninstalled, the test for required packages should fail.

#### E2E Tests "{new-project}/src/tests/e2e"
- The new project should include a `nightwatch.conf.js` file.
- It should also include a `.env` and a `.gitignore` file. 
- Running `npm run test:e2e` should install the required node packages if not installed previously.
- It should run the tests on the saved link value `E2E_URL` in the `.env` file.
- If the saved link is not up and running, the tests should provide appropriate failure feedback.
- Running `npm run test:e2e` should run the localization e2e test with appropriate CLI feedback.
- It should run the search e2e test with appropriate CLI feedback.
- The command `npm run test:e2e` should also be able to run with the `--url` flag, where the users can pass a custom URL to test. Example usage = `npm run test:e2e --url=https://www.ontario.ca/international-trade-events-calendar`.

### Next step
- Should make the integration tests work with the updated add and remove package commands. If the user adds eslint or prettier after the project is created, the variables in `src/tests/integration/packages.spec.js` should update accordingly, or the user should be provide appropriate feedback to update those variables.

### CODE REVIEW

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/package.njk [hunk 1]
```diff
@@ -6,7 +6,8 @@
   "scripts": {
     "build": "rm -rf dist && eleventy",
     "debug": "rm -rf dist && DEBUG=* eleventy",
-    "serve": "eleventy --serve"{% if addESLint %},
+    "serve": "eleventy --serve",
+    "test": "npm list mocha chai || npm install mocha chai@4.4.1; mocha './src/tests/**/*.spec.js'"{% if addESLint %},
```
Comment (syed-ods) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/package.njk]: Need to fix the `chai` version and fix the ES_Module error that stems from the latest `chai` package.

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/gitignore.njk [hunk 2]
```diff
@@ -0,0 +1,6 @@
+.env
+dist/
+logs/
+node_modules/
+tests_output/
+DS_Store
```
Comment (kyle-binger-3) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/gitignore.njk]: Update this to `**/.DS_Store` to make sure that `.DS_Store` files generated within any directory are ignored. Just a further safety net.

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/localization.spec.njk [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+ * @param {string} langAttr - Expected value of the HTML lang attribute
+ * @param {string} buttonText - Expected text content of the language toggle button
+ */
+function checkLanguageAttributes(langAttr, buttonText) {
+  browser.assert
+    .attributeEquals("html", "lang", langAttr) // Check if the HTML lang attribute matches the expected value
+    .getText(".ontario-header__language-toggler span", function (result) {
+      this.assert.equal(result.value.trim(), buttonText); // Check if the button text matches the expected value
+    });
+}
```
Comment (kyle-binger-3) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/localization.spec.njk]: This is great but I wonder if there's value in making it more reusable like you did with `checkElement()` above?

Perhaps breaking it into two separate functions/tests? e.g.

`checkElementAttributes(selector, attributes)` and
`checkButtonText(selector, expectedText)`

If you are soley using this just for checking the lang button then it may be over-engineering, however if you are doing other element checks like this it makes things more reusable.

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/nightwatch.conf.njk [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  },
+
+  test_settings: {
+    default: {
+      desiredCapabilities: {
+        browserName: "chrome",
+      },
+    },
+  },
+};
```
Comment (kyle-binger-3) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/nightwatch.conf.njk]: I know very little about nightwatch. I'm just wondering, is there value in doing tests in Microsoft Edge as well? Might be a future state thing, Chrome and Edge are the OPS supported browsers.

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/search.spec.njk [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+/**
+ * Helper function to check if element exists and is visible
+ * @param {string} selector - Selector value, either a class or an id, of the element
+ * @param {string} description - Description of the element
+ */
+function checkElement(selector, description) {
+  browser.assert
+    .elementPresent(selector, `${description} exists`)
+    .assert.visible(selector, `${description} is visible`);
+}
```
Comment (kyle-binger-3) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/search.spec.njk]: Do these functions have to be redefined per test? Or can they be created in a "Tests Util" folder and imported for use?

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  });
+
+  // The optional packages and their associated packages
+  const optionalPackages = {
+    eslint: {
+      enabled: {{ addESLint }},
+      packages: [
+        "eslint",
+        "eslint-plugin-import",
+        "@ongov/eslint-config-ontario-frontend",
```
Comment (kyle-binger-3) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk]: I think it's worth importing the config objects for ESLint and Prettier and looking for them under the `packages` key. Keeps everything referencing a single source of truth.

You can find the config files under `ontario-frontend-cli/core/config/packages`. Within `index.js` they get combined and exported as `PACKAGES_CONFIG`.

Comment (kyle-binger-3) [reply packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk]: Realizing that since these test files get copied from the cli during project generation that this might be tricky. Perhaps a tests global variable file that makes a copy of the packageConfig `{...PACKAGE_CONFIG}` would retain it upon copy allowing it to be available for use? Worth exploring.

**Review by kyle-binger-3 — CHANGES_REQUESTED** [review]
Hey Syed!

This is fantastic work, great job capturing all the different kinds of tests (unit, E2E, Integration).

I've left some general feedback that may be helpful. Let me know if you have any questions / want to chat further.

Thanks!

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/prompt.spec.njk
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/prompt.spec.njk:1]: The prompts provide clear instructions for running the E2E tests, well done. 

Consider leveraging the `textStyling.js` util and the `chalkColours.js` util.
 
Consider including additional information about setting up environment variables and troubleshooting common issues.

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/localization.spec.njk
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/localization.spec.njk:1]: This file is well-structured, and the use of helper functions makes it modular and easy to read.

Consider additional comments that explain various blocks of code.

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/localization.spec.njk [hunk 7]
```diff
@@ -0,0 +1,77 @@
+{# Localization End to End tests to check if the basic language toggle between EN and FR is working #}/**
+ * E2E tests that check if the localization content is working, including the language toggle button.
+ */
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/localization.spec.njk]:
```suggestion
/**
 * E2E tests that check if the localization content is working, including the language toggle button.
 * This suite ensures that the language toggles between English and French correctly.
 */
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/localization.spec.njk [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * Helper function to check if element exists and is visible
+ * @param {string} selector - Selector value, either a class or an id, of the element
+ * @param {string} description - Description of the element
+ */
+function checkElement(selector, description) {
+  browser.assert
+    .elementPresent(selector, `${description} exists`)
+    .assert.visible(selector, `${description} is visible`);
+}
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/localization.spec.njk]: Could consider adding a timed check to identify render delays:

`.waitForElementVisible(selector, 1000, `${description} is visible within 1 second`);`

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/localization.spec.njk [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * Helper function to check the language attributes and button text
+ * @param {string} langAttr - Expected value of the HTML lang attribute
+ * @param {string} buttonText - Expected text content of the language toggle button
+ */
+function checkLanguageAttributes(langAttr, buttonText) {
+  browser.assert
+    .attributeEquals("html", "lang", langAttr) // Check if the HTML lang attribute matches the expected value
+    .getText(".ontario-header__language-toggler span", function (result) {
+      this.assert.equal(result.value.trim(), buttonText); // Check if the button text matches the expected value
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/localization.spec.njk]: Please add some error messages in your assertions to make debugging easier.

```js
.attributeEquals("html", "lang", langAttr, `Expected lang attribute to be ${langAttr}`);
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/localization.spec.njk [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+ * @param {string} frLangAttr - The French language HTML attribute value.
+ * @param {string} frButtonText - The language toggle button value to French.
+ * @param {string} enButtonText - The language toggle button value to English.
+ */
+
+module.exports = {
+  before: function (browser) {
+    // Actions to perform before the test suite runs
+    browser.maximizeWindow();
+  },
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/localization.spec.njk]: Good use of `before` hook here. 

Suggestion: Ad a check to ensure `E2E_URL` is defined and accessible:

```js
if (!E2E_URL) {
    throw new Error("E2E_URL is not defined. Please set the environment variable.");
}
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/localization.spec.njk [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    // Initial check for English
+    checkLanguageAttributes(enLangAttr, frButtonText); // Check English attributes and French button text
+
+    // Toggle to French and check
+    browser.click(langButton); // Click the language toggle button to switch to French
+    checkLanguageAttributes(frLangAttr, enButtonText); // Check French attributes and English button text
+
+    // Toggle back to English and check
+    browser.click(langButton); // Click the language toggle button to switch back to English
+    checkLanguageAttributes(enLangAttr, frButtonText); // Check English attributes and French button text
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/localization.spec.njk]: The checks for toggling between languages are well-implemented. 

Consider adding a short wait between clicks to ensure the page has time to update:

```js
browser.pause(500); // Wait 0.5 seconds before checking the next language state
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/localization.spec.njk
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/localization.spec.njk:1]: There are some edge cases you could test:
- Verify the default language of the page when it loads
- The language toggle is not spammable, or at least performs properly.

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/nightwatch.conf.njk
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/nightwatch.conf.njk:1]: The configuration is straightforward and clear, focusing on Chrome for end-to-end tests. Well done!

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/nightwatch.conf.njk [hunk 12]
```diff
@@ -0,0 +1,17 @@
+module.exports = {
+  src_folders: ["src/tests/e2e/**/**.spec.js"],
+
+  webdriver: {
+    start_process: true,
+    server_path: require("chromedriver").path,
+    port: 9515,
+  },
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/nightwatch.conf.njk]: Good setup with chromedriver. Ensure the chromedriver version is compatible with the installed Chrome version.

Idea: Add an environment variable for server_path to easily switch drivers if needed.
```js
server_path: process.env.CHROMEDRIVER_PATH || require("chromedriver").path,
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/nightwatch.conf.njk [hunk 13]
```diff
@@ -0,0 +1,17 @@
+module.exports = {
+  src_folders: ["src/tests/e2e/**/**.spec.js"],
+
+  webdriver: {
+    start_process: true,
+    server_path: require("chromedriver").path,
+    port: 9515,
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/nightwatch.conf.njk:7]: Consider making this more configurable via env var:

```js
port: process.env.WEBDRIVER_PORT || 9515,
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/nightwatch.conf.njk (hunk 4 above)
Comment (matt-stjean) [reply packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/nightwatch.conf.njk]: Example based on Kyle's comment.

```js
test_settings: {
  default: {
    desiredCapabilities: {
      browserName: "chrome",
    },
  },
  firefox: {
    desiredCapabilities: {
      browserName: "firefox",
    },
  },
  edge: {
    desiredCapabilities: {
      browserName: "MicrosoftEdge",
      'ms:edgeOptions': {
        args: ['--headless', '--disable-gpu']
      }
    },
  },
},
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/nightwatch.conf.njk
Comment (matt-stjean) [reply packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/nightwatch.conf.njk:1]: Consider adding a global timeout setting for slow environments:
```js
globals: {
  waitForConditionTimeout: 5000, // 5 seconds
},
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/prompt.spec.njk [hunk 14]
```diff
@@ -0,0 +1,15 @@
+{# CLI prompts for the user when they run the command npm test:e2e to run our end to end tests #}/**
+ * E2E test prompts to give user feedback about the command structure and how to run the tets.
+ */
+let E2E_URL = process.env.E2E_URL
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/prompt.spec.njk]: Ensure the `E2E_URL` env var is defined with a conditional check:

```js
if (!E2E_URL) {
    console.error(`${fgYellow}E2E_URL is not defined. Please set the environment variable in your .env file.${reset}`);
    process.exit(1);
}
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/prompt.spec.njk [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+
+const reset = "\x1b[0m";
+const fgCyan = "\x1b[36m";
+const fgYellow = "\x1b[33m";
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/prompt.spec.njk]: Use `Chalk.js` which is installed already and available through some utils.

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/prompt.spec.njk [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+
+console.log(`\n${fgCyan}Make sure to run the project locally or have a working URL before running this test.${reset}`);
+console.log(`E2E tests run on ${fgCyan}Google Chrome${reset}.`);
+console.log(`The E2E tests will run on the saved URL in the environment variables ${fgCyan}E2E_URL${reset}=${fgCyan}${E2E_URL}${reset}`);
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/prompt.spec.njk]: Maybe add a desc of what the tests are doing:

```js
console.log(`E2E tests will run automated checks to ensure the application's functionality is working as expected.`);
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/prompt.spec.njk [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+console.log(`You can update the ${fgCyan}E2E_URL${reset} in your local .env file`);
+console.log(`To run the tests on a target URL, use the command: ${fgYellow}npm run test:e2e ${fgCyan}--url=(Your-Site-URL)${reset}`);
+console.log(`Example: ${fgYellow}npm run test:e2e${reset} ${fgCyan}--url=https://www.ontario.ca/international-trade-events-calendar${reset}\n`);
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/prompt.spec.njk]: Maybe add a note that they should replace the placeholder url

```js
console.log(`(Replace the placeholder URL with the actual URL you want to test)`);
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/search.spec.njk [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+  "E2E: Search Bar Test": function (browser) {
+    const E2E_URL = process.env.E2E_URL;
+    const searchTerm = "license";
+    const expectedSearchURL =
+      "https://www.ontario.ca/search/search-results/?query=license";
+    const searchInputSelector = "#ontario-search-input-field";
+    const searchButtonSelector = "#ontario-search-submit";
+
+    browser
+      .url(E2E_URL)
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/search.spec.njk]:
```js
if (!E2E_URL) {
    console.error('E2E_URL is not defined. Please set the environment variable in your .env file.');
    process.exit(1);
}
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/search.spec.njk (hunk 5 above)
Comment (matt-stjean) [reply packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/search.spec.njk]: Nitpicky on the messaging:
```js
.elementPresent(selector, `${description} should be present`)
.assert.visible(selector, `${description} should be visible`);
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/search.spec.njk [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+      .perform(() => checkElement(searchInputSelector, "Search input field"))
+      .perform(() => checkElement(searchButtonSelector, "Search button"))
+
+      // Type in the search term and check redirection
+      .setValue(searchInputSelector, searchTerm)
+      .click(searchButtonSelector)
+      .waitForElementVisible("body", 2000, "Body is visible after search")
+      .assert.urlContains(
+        expectedSearchURL,
+        "Redirected to the correct search URL"
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/search.spec.njk]: `.assert.urlContains(expectedSearchURL, 'Should redirect to the correct search URL')`

Could use clearer language

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/search.spec.njk [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+ *
+ * Variables and parameters for the e2e tests
+ * @param {string} E2E_URL - The saved env variable path of the link to be tested, local, staging, or prod.
+ * @param {string} searchTerm - The search text to be typed in the search box.
+ * @param {string} expectedSearchURL - The URL that the search term is supposed to be redirected to.
+ * @param {string} searchInputSelector - The selector for the search field.
+ * @param {string} expectedSearchURL - The selector for the search submit button.
+ */
+
+module.exports = {
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/search.spec.njk]: We could have some edge cases covered?
```js
'E2E: Empty Search Term': function (browser) {
    browser
        .url(E2E_URL)
        .waitForElementVisible('body', 2000)
        .perform(() => checkElement(searchInputSelector, 'Search input field'))
        .perform(() => checkElement(searchButtonSelector, 'Search button'))
        .click(searchButtonSelector)
        .assert.containsText('#error-message', 'Please enter a search term.')
        .end();
}
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk:1]: The integration tests are well-structured, and the use of helper functions enhances readability. Good job on ensuring coverage for both required and optional packages.

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+      } catch (e) {
+        expect.fail(`Failed: to detect the required package: "${pkg}"`);
+      }
+    });
+  });
+
+  // The optional packages and their associated packages
+  const optionalPackages = {
+    eslint: {
+      enabled: {{ addESLint }},
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk]: Ensure CLI prompts correctly set these values before running the tests. Add a check that confirms the values are defined, example:

```js
if (typeof addESLint === 'undefined' || typeof addPrettier === 'undefined') {
    console.error('addESLint or addPrettier is not defined. Please ensure the CLI prompts set these values correctly.');
    process.exit(1);
}
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+ * The boolean value should be captured from the CLI prompts.
+ * 
+ * @returns {boolean} Based on the test, the result will either be TRUE or FALSE (error).
+ */
+
+// A utility function to read package.json
+const readPackageJson = () => {
+  const packageJsonPath = path.resolve(__dirname, "../../..", "package.json");
+  return JSON.parse(fs.readFileSync(packageJsonPath, "utf8"));
+};
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk]: Do we need to make this utility?

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+// Checks if a package exists in package.json both in dependencies and dev dependencies
+const isPackageInstalled = (pkg) => {
+  const packageJson = readPackageJson();
+  return (
+    (packageJson.dependencies &&
+      packageJson.dependencies.hasOwnProperty(pkg)) ||
+    (packageJson.devDependencies &&
+      packageJson.devDependencies.hasOwnProperty(pkg))
+  );
+};
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk]: Are we not able to leverage the existing utility methods?

Comment (matt-stjean) [reply packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk]: This code is pretty clean but could be a bit clearer

```js
const isPackageInstalled = (pkg) => {
    const packageJson = readPackageJson();
    if (packageJson.dependencies && packageJson.dependencies.hasOwnProperty(pkg)) {
        console.log(`Package ${pkg} found in dependencies`);
        return true;
    }
    if (packageJson.devDependencies && packageJson.devDependencies.hasOwnProperty(pkg)) {
        console.log(`Package ${pkg} found in devDependencies`);
        return true;
    }
    return false;
};
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+    it(`The core package "${pkg}" should be installed`, function () {
+      try {
+        if (!isPackageInstalled(pkg)) {
+          throw new Error();
+        }
+      } catch (e) {
+        expect.fail(`Failed: to detect the required package: "${pkg}"`);
+      }
+    });
+  });
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk]: Good use of loops to test optional packages. This could leverage more modern JS and add better traceability, and we should be ensuring that 'enabled' flags are correctly set by CLI prompts:

```suggestion
for (const [parentPackage, config] of Object.entries(optionalPackages)) {
    if (config.enabled) {
        console.log(`Testing optional package: ${parentPackage}`);
        config.packages.forEach((pkg) => {
            it(`The package "${pkg}" should be installed for "${parentPackage}"`, function () {
                try {
                    if (!isPackageInstalled(pkg)) {
                        throw new Error();
                    }
                } catch (e) {
                    expect.fail(`Failed: couldn't locate the package: "${pkg}" for "${parentPackage}"`);
                }
            });
        });
    }
}
```

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/unit/app.spec.njk
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/unit/app.spec.njk:1]: The unit tests are well-structured and comprehensive, covering critical file existence checks. Good use of helper functions to maintain clean code.

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/unit/app.spec.njk [hunk 25]
```diff
@@ -0,0 +1,90 @@
...
+ * @returns {boolean} TRUE/FALSE value based on if the file was found or
+ */
+function checkFileExists(filePath, description, customErrorMessage) {
+  it(description, function () {
+    assert(
+      existsSync(filePath),
+      `File does not exist: ${filePath} \n ${customErrorMessage}`,
+    );
+  });
+}
```
Comment (matt-stjean) [inline packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/unit/app.spec.njk]: Do we need to remake this util?

**Review by matt-stjean — CHANGES_REQUESTED** [review]
Great work on implementing the tests and configurations! The structure is clear, and the use of helper functions enhances readability and maintainability. Here are some key points and suggestions:

1. **Environment Variables**: Ensure all required environment variables are set and provide clear error messages if they are not.
2. **Helper Functions**: Improve error handling and add detailed logs to assist with debugging.
3. **Test Cases**: Add edge cases and more comprehensive checks to ensure robustness.
4. **Documentation**: Update comments and instructions to guide future updates and changes.

Overall, your implementation is solid, with minor enhancements needed for error handling and edge cases. Keep up the good work!

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk (hunk 6 above)
Comment (syed-ods) [reply packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk]: I believe I would need to add scripts to make this data object accessible to the templates.
@matt-stjean, should I invest some time in that direction?

Comment (matt-stjean) [reply packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/integration/packages.spec.njk]: @syed-ods I think it's worth investing some time into it. Give yourself a time box, giving yourself a time limit of how much you think you should spend on it. When your time is up, we can meet up and talk it through?

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/prompt.spec.njk
Comment (syed-ods) [reply packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/prompt.spec.njk:1]: @matt-stjean in order to export colours and make them usable in a newly created project. I would not only need to export the colours' config, but also install extra packages to make it work. Should I use similar hard coded ASCII colours, or install the chalk packages on user project and then utilize our config colours?

Comment (matt-stjean) [reply packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/prompt.spec.njk:1]: I forgot that this was in newly created projects, and not in our monorepo. **Just hardcode it.** 

We could expose it through Ontario-Frontend dependency, and makes me realize we could easily expose our logger and other things too this way, but alas our ship is sinking.

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/package.njk (hunk 1 above)
Comment (syed-ods) [reply packages/ontario-frontend-cli/resources/templates/ontario-create-app/package.njk]: https://github.com/chaijs/chai/issues/1561

File: packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/nightwatch.conf.njk (hunk 13 above)
Comment (syed-ods) [reply packages/ontario-frontend-cli/resources/templates/ontario-create-app/tests/e2e/nightwatch.conf.njk:7]: I tried using the above logic `process.env.WEBDRIVER_PORT || 9515` but its only able to use a static value, the either or and if logic is not working within the config file or in the webdriver.port value.

**Comment by syed-ods** [comment]
@matt-stjean I'm thinking of making Microsoft Edge as the default browsers for e2e tests. That's because I tried uninstalling Chrome and run the tests but Nightwatch was unable to run the tests. I also tried installing specific webdrivers for each browser. It seems like an MVP idea to keep one browser only and mention the details in our developer documentation.

**Comment by matt-stjean** [comment]
Chrome is installed on every OPS machine, though, I don’t think we need to handle a case where Chrome is not installed.

**Review by matt-stjean — APPROVED** [review]
LGTM, thanks for your work Syed
